gpytorch/lazy/kronecker_product_added_diag_lazy_tensor.py

```diff
@@ -0,0 +1,139 @@
+                    "DiagKronProductLazyTensor expects lazy tensors with the "
+                    "same batch shapes. Got {}.".format([lv.batch_shape for lv in lazy_tensors])
+                )
+        super(_DiagKroneckerProdLazyTensor, self).__init__(*lazy_tensors)
+        self.lazy_tensors = lazy_tensors
+
+    def get_diag(self):
+        sz1 = self.lazy_tensors[0].size(0)
+        sz2 = self.lazy_tensors[1].size(0)
+
+        d1 = self.lazy_tensors[0].diag()
+        d2 = self.lazy_tensors[1].diag()
+
+        out = d2.expand(sz1, sz2).t().mul(d1)
+
+        return out.t().contiguous().view(sz1*sz2)
+
+
+class _KroneckerProductLazyLogDet(KroneckerProductLazyTensor):
+    def __init__(self, *lazy_tensors, jitter=settings.tridiagonal_jitter.value()):
+        super(_KroneckerProductLazyLogDet, self).__init__(*lazy_tensors)
+        # on initialization take the eigenvectors & eigenvalues of all of the lazy tensors
+        self.eig_cache = [torch.symeig(lt.evaluate(), eigenvectors = True) for lt in self.lazy_tensors]
+
+    def inv_matmul(self, rhs, jitter=settings.tridiagonal_jitter.value()):
+        Vinv = KroneckerProductLazyTensor(*[DiagLazyTensor(1 / (s[0].abs()+jitter) ) for s in self.eig_cache])
+        Q = KroneckerProductLazyTensor(*[NonLazyTensor(s[1]) for s in self.eig_cache])
+
+        # first compute Q^T y
+        res1 = Q.t().matmul(rhs)
+
+        # now V^{-1} Q^T y
+        res2 = Vinv.matmul(res1)
+        res3 = Q.matmul(res2)
+
+        return res3
+
+    def logdet(self):
+        lt_sizes = [lt.size(-1) for lt in self.lazy_tensors]
+
+        # det(A \kron B) = det(A)^m det(B)^n where m,n are the sizes of A,B
+        scaled_logdets = [m * s[0].sum() for m, s in zip(lt_sizes, self.eig_cache)]
+
+        full_logdet = 0.
+        for logdet in scaled_logdets:
+            full_logdet = logdet + full_logdet
+        
+        return full_logdet
+
+
+class KroneckerProductPlusDiagLazyTensor(AddedDiagLazyTensor):
+    def __init__(self, *lazy_tensors):
+        super(KroneckerProductPlusDiagLazyTensor, self).__init__(*lazy_tensors)
+        if len(lazy_tensors) > 2:
+            raise RuntimeError("An AddedDiagLazyTensor can only have two components")
+        if isinstance(lazy_tensors[0], DiagLazyTensor) and isinstance(lazy_tensors[1], DiagLazyTensor):
+            raise RuntimeError("Trying to lazily add two DiagLazyTensors. " "Create a single DiagLazyTensor instead.")
+        elif isinstance(lazy_tensors[0], DiagLazyTensor):
+            self._diag_tensor = lazy_tensors[0]
+            self._lazy_tensor = lazy_tensors[1]
+        elif isinstance(lazy_tensors[1], DiagLazyTensor):
+            self._diag_tensor = lazy_tensors[1]
+            self._lazy_tensor = lazy_tensors[0]
+        else:
+            raise RuntimeError("One of the LazyTensors input to AddedDiagLazyTensor must be a DiagLazyTensor!")
+
+    def inv_quad_logdet(self, inv_quad_rhs=None, logdet=False, reduce_inv_quad=True):
```

Comment:
Is this necessary? Seems like the default implementation.

Reply:
well the default implementation will use CG for the solves for large tensors, which we'd like to avoid here